lib/active_model/serializer/adapter/cached_serializer.rb

```diff
@@ -27,17 +29,21 @@ def fragment_cached?
           @klass._cache_only && !@klass._cache_except || !@klass._cache_only && @klass._cache_except
         end
 
-        def cache_key
+        def cache_key(adapter_instance)
           parts = []
           parts << object_cache_key
+          parts << adapter_instance.class.to_s.demodulize.underscore
           parts << @klass._cache_digest unless @klass._cache_options && @klass._cache_options[:skip_digest]
           parts.join('/')
         end
 
         def object_cache_key
+          return @cached_serializer.object.cache_key if @cached_serializer.object.respond_to? :cache_key
+
+          fail(UndefinedCacheKey, "#{@cached_serializer.object} must define #cache_key, or the cache_key option must be passed into cache on #{@cached_serializer}") if @klass._cache_key.blank?
           object_time_safe = @cached_serializer.object.updated_at
           object_time_safe = object_time_safe.strftime('%Y%m%d%H%M%S%9N') if object_time_safe.respond_to?(:strftime)
-          (@klass._cache_key) ? "#{@klass._cache_key}/#{@cached_serializer.object.id}-#{object_time_safe}" : @cached_serializer.object.cache_key
+          "#{@klass._cache_key}/#{@cached_serializer.object.id}-#{object_time_safe}"
```

Comment:
Can we move the `cache_key` method into the serializer? Would simplify things I think

[Serializer::Caching](https://github.com/rails-api/active_model_serializers/blob/7d4f0c5c8af93c32bd69ed36b2433de42243f93c/lib/active_model/serializer/caching.rb)

```ruby
module ActiveModel
  class Serializer
    module Caching
      extend ActiveSupport::Concern

      # Use object's cache_key if available, else derive a key from the object
      # Pass the `key` option to the `cache` declaration or override this method to customize the cache key
      def cache_key
        if object.respond_to?(:cache_key)
          object.cache_key
        else
          object_time_safe = object.updated_at
          object_time_safe = object_time_safe.strftime('%Y%m%d%H%M%S%9N') if object_time_safe.respond_to?(:strftime)
          "#{self.class._cache_key}/#{object.id}-#{object_time_safe}"
        end
      end
    end
  end
end
```